implementations/rust/ockam/ockam_identity/src/secure_channel/decryptor.rs

```diff
@@ -1,39 +1,137 @@
 use crate::identity::IdentityError;
-use crate::secure_channel::encryptor::Encryptor;
+use crate::secure_channel::encryptor::{Encryptor, KEY_RENEWAL_INTERVAL};
 use ockam_core::compat::sync::Arc;
 use ockam_core::compat::vec::Vec;
-use ockam_core::vault::{KeyId, SymmetricVault};
+use ockam_core::vault::KeyId;
 use ockam_core::Result;
+use ockam_key_exchange_xx::XXInitializedVault;
+use tracing::warn;
 
-#[derive(Clone)]
 pub(crate) struct Decryptor {
-    key: KeyId,
-    vault: Arc<dyn SymmetricVault>,
+    current_key: KeyId,
+    current_key_nonce: u64,
+
+    previous_key: Option<KeyId>,
+    vault: Arc<dyn XXInitializedVault>,
+    nonce_tracker: NonceTracker,
+}
+
+pub(crate) struct NonceTracker {
+    nonce_bitmap: u16,
```

Comment:
I don't get this part, could you please elaborate? We have KEY_RENEWAL_INTERVAL = 10. Imagine Secure Channel was just created, we can accept nonces from 0 to 19 inclusively. How can we track which values were already used with just 16 bits of state?

Reply:
> we can accept nonces from 0 to 19 inclusively
True, but we are keeping track only of the last 10, when an increased nonce is received we move the window to that nonce.

Reply:
I'm a bit concerned on the dependency of a specific size.    I mean,   the 10 value was taked out of thin air just as a number to put there and check things works,   but at least until getting a :heavy_check_mark:  that we are going to stick to that number,   I'd have expected the code to be ok to change to  20,  50, 100, whatever by changing the constant. 

The out-of-order window doesn't affect interoperability,   just how much tolerant to out-of-order messages a receiver is.. so I think this is not a big blocker  (I mean,  if we decide to rekey each 100 messages,   and this code still works only that with a nonce validity window of 10,   i would say its ok.  And we can work on how to make this window modifiable in some future work).

Reply:
Maybe we could use https://docs.rs/bitmaps/latest/bitmaps/ to make things cleaner?

Reply:
I'm not convinced introducing an extra dependency for a simple bitmap is worth the trouble, the implementation works as long `KEY_RENEWAL_INTERVAL` is smaller than the number of bits in the `nonce_bitmap` (this would be true even if we were using the library). I'm adding a check for it, it fails at compile time if that's not the case.
It should work up to 128 messages, after that, we would probably need to switch to another implementation.